pkg/workload/histogram/exporter/util.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package exporter
+
+import (
+	"regexp"
+	"time"
+
+	"github.com/codahale/hdrhistogram"
+	"github.com/gogo/protobuf/proto"
+	prom "github.com/prometheus/client_model/go"
+)
+
+func sanitizeOpenmetricsLabels(input string) string {
+	regex := regexp.MustCompile(`[^a-zA-Z0-9_]`)
+	sanitized := regex.ReplaceAllString(input, "_")
+	firstCharRegex := regexp.MustCompile(`^[^a-zA-Z_]`)
+	sanitized = firstCharRegex.ReplaceAllString(sanitized, "_")
+	return sanitized
+}
+
+// ConvertHdrHistogramToPrometheusMetricFamily converts a Hdr histogram into MetricFamily which is used
+// by expfmt.MetricFamilyToOpenMetrics to export openmetrics
+func ConvertHdrHistogramToPrometheusMetricFamily(
+	h *hdrhistogram.Histogram, name *string, start time.Time, labels *map[string]string,
+) *prom.MetricFamily {
+	hist := &prom.Histogram{}
+
+	bars := h.Distribution()
+	hist.Bucket = make([]*prom.Bucket, 0, len(bars))
+
+	var cumCount uint64
+	var sum float64
+	for _, bar := range bars {
+		if bar.Count == 0 {
+			// No need to expose trivial buckets.
+			continue
+		}
+		upperBound := float64(bar.To)
+		sum += upperBound * float64(bar.Count)
```

Comment:
This can overshoot wildly. Ideally, we should be able to compute an _exact_ sum of observations. The hdrhistogram does expose the raw observations (`Snapshot.Counts`), so it's technically possible with the help of `ValueAtQuantile`. I also wonder if we should instead use static buckets.

Reply:
Do take a look at `newHistogram` (in `pkg/util/metric/metric.go`), and the corresponding static buckets in `histogram_buckets.go`. Maybe `IOLatencyBuckets` would suffice for most (all?) performance tests.

Reply:
Actually that's a good point, we definitely can use static buckets. But then our openmetrics histogram won't be directly comparable with that of the current `stats.json` one.

Reply:
`IOLatencyBucket` looks good for our usecases, I'll check i there's any case of more being used

Reply:
@srosenberg, `hdrhistogram` doesn't provide good APIs to provide total sum. The closest approximation (which might be better than ours) is to get the `Mean` and multiply it by the `TotalCount`. I don't think we will be able to get more closer than this approach of getting the sum.
Other approach might be to skip sum altogether to avoid incosistent metric sum.

Which approach do you think is the best?

Reply:
We synced up offline and we came to a decision to emit `summary` metric and not save `sum` altogether. `sum` equal to zero as undefined to avoid emitting a wrong metric.